test/fixtures/iotFleetProvisioning/registerDevice.js

```diff
@@ -0,0 +1,133 @@
+'use strict';
+// NOTE: `aws-iot-device-sdk-v2` is bundled into the deployment package
+// eslint-disable-next-line import/no-unresolved
+const { mqtt, io, iot, iotidentity } = require('aws-iot-device-sdk-v2');
+
+const subscribe = (fn, channelName, isRejectionChannel) => {
+  let subscription;
+  const message = new Promise((resolve, reject) => {
+    subscription = fn((error, response) => {
+      if (response) {
+        console.log(`${channelName} message: ${JSON.stringify(response)}`);
+      }
+
+      if (error || !response) {
+        console.log(`Error occurred on channel ${channelName}.`);
+        reject(error);
+      } else {
+        isRejectionChannel ? reject(response) : resolve(response);
+      }
+    });
+  });
+  return { subscription, message };
+};
+
+const executeKeys = async identity => {
+  console.log('Subscribing to CreateKeysAndCertificate Accepted and Rejected topics..');
+
+  const keysSubRequest = {};
+
+  const createKeysAndCertificateAcceptedChannel = subscribe(
+    identity.subscribeToCreateKeysAndCertificateAccepted.bind(
+      identity,
+      keysSubRequest,
+      mqtt.QoS.AtLeastOnce
+    ),
+    'createKeysAndCertificateAccepted',
+    false
+  );
+  const createKeysAndCertificateRejectedChannel = subscribe(
+    identity.subscribeToCreateKeysAndCertificateRejected.bind(
+      identity,
+      keysSubRequest,
+      mqtt.QoS.AtLeastOnce
+    ),
+    'createKeysAndCertificateRejected',
+    true
+  );
+
+  await Promise.all([
+    createKeysAndCertificateAcceptedChannel.subscription,
+    createKeysAndCertificateRejectedChannel.subscription,
+  ]);
+
+  console.log('Publishing to CreateKeysAndCertificate topic..');
+  const keysRequest = {
+    toJSON() {
+      return {};
+    },
+  };
+
+  await identity.publishCreateKeysAndCertificate(keysRequest, mqtt.QoS.AtLeastOnce);
+
+  return Promise.race([
+    createKeysAndCertificateAcceptedChannel.message,
+    createKeysAndCertificateRejectedChannel.message,
+  ]);
```

Comment:
Not sure I understand, race allows me to resolve or reject as soon as one promise resolve or reject. You think it's not a good idea to create a promise (for the rejected channel) that'll never resolve or reject because no message will actually hit the MQTT channel ? What are the consequences of creating such Promise ?

I'll implement your solution anyway, just wanted to understand the logic behind not creating promises that will not resolve :)

Reply:
> What are the consequences of creating such Promise ?

It's more about cleanness of a flow setup. Creating a promise for _async_ action that we know will never happen feels dirty and also blurs the picture for other developers that may want to follow the internal logic.

Otherwise obviously there's no harm.